new package: kitty terminal

SteVwonder opened the pull request:
Adds a package for the kitty terminal.

A few of the dependencies needed `setup_dependent_environment` implemented in order to be detected/used. For example, the `util-macros` package provides several m4 macros, which aren't discovered unless `ACLOCAL_PATH` is properly set.

One of kitty's dependencies, `libxkbcommon`, needed a few dependencies of its own added in order to compile.

adamjstewart commented on var/spack/repos/builtin/packages/kitty/package.py:
```diff
@@ -0,0 +1,55 @@
+    version('0.10.1', sha256='ef22208497a76e2f88ebe56c176e4608f049b056252cf1bf122c9c1ec711cfa6')
+    version('0.10.0', sha256='056563862c5759b740e95efff44b82c1a4efc370092f22f26aee0b774106bf4d')
+
+    depends_on('python@3.7:', type=('build', 'run'))
```
According to https://sw.kovidgoyal.net/kitty/build.html, it only needs 3.5 or newer.

SteVwonder replied:
Good catch! Updated.

adamjstewart commented on var/spack/repos/builtin/packages/kitty/package.py:
```diff
@@ -0,0 +1,55 @@
+    def install(self, spec, prefix):
+        with working_dir(self.build_directory):
+            self.python('-s', 'setup.py', 'linux-package',
+                        '--prefix={0}'.format(prefix))
```
It might be easier to override `install_args` instead.

SteVwonder replied:
Unfortunately, the `setup.py` in `kitty` in non-standard in two ways:

1. it requires a target of `linux-package` rather than `install`
2. it barfs on the `--no-user-cfg` flag that is added by `PythonPackage`'s `setup_py` (which is called by `install`) 

So this avoids both problems.  Number 1 can be avoided by having `self.setup_py('linux-package')` as the `install` method, but then number 2 is still a problem.  Looking at `setup_py`, I didn't see a way to avoid having `--no-user-cfg` added to the command.

hartzell replied:
This seems novel enough that it would be worth adding a comment to the install method that summarizes the reasoning for the next person to touch it.

SteVwonder replied:
:+1: @hartzell: in my last commit, I added a comment to the `install` method in the `kitty` package.  For whatever reason, GitHub is not rendering it in this preview (it is still showing the code from back in October).  You can see it in the full diff page. Pasted below for convenience:
```python
    def install(self, spec, prefix):
        # kitty's setup.py does not recognize the '--no-user-cfg' flag that is
        # used by default in the setup_py method, overriding that behavior here
```

alalazo commented:
@SteVwonder

SteVwonder commented:
:scream: Wow, I really let this one linger.  Sorry guys!

Rebased on the upstream `develop` and addressed the python dependency.  Please see my GitHub and source code comments about the `install` method.

SteVwonder commented:
Bump

SteVwonder commented:
Ping for merging :smiley:

@alalazo?

alalazo commented on var/spack/repos/builtin/packages/libxkbcommon/package.py:
```diff
@@ -18,8 +18,13 @@ class Libxkbcommon(AutotoolsPackage):
+    depends_on('autoconf', type='build')
+    depends_on('automake', type='build')
+    depends_on('libtool', type='build')
+    depends_on('m4', type='build')
```
Why have these dependencies been added? Is there a missing `configure` somewhere?

SteVwonder replied:
Good call.  I'm not sure why I added those.  Removed and build seems to happen just fine without them.  I also confirmed that the tarballs of both versions of libxkbcommon have `configure` scripts.

While making that change, I also rebase'd onto the latest upstream `develop`.
